camply/providers/going_to_camp/going_to_camp_provider.py

```diff
@@ -383,6 +384,8 @@ def _hostname_for(self, recreation_area_id: int) -> str:
                 return hostname
         return None
 
+    @ratelimit.sleep_and_retry
+    @ratelimit.limits(calls=3, period=1)
     def _api_request(
```

Comment:
Hey @acaloiaro what do you think of this change? It would limit making more than 3 calls to the GoingToCamp API in a single second. 

I've been using this command as a baseline:
```
camply --debug campsites \
  --provider goingtocamp \
  --rec-area 4 \
  --campground -2147483645 \
  --start-date 2023-09-01 \
  --end-date 2023-09-02
  ```

I've seen ratelimiting about double that command runtime from 9 -> 18s in my testing. How do you feel about that performance hit?

Reply:
That doubling in time is a bit of a bummer. 

Were you being rate limited, or did someone report being rate limited? If so, the jump in request time is certainly better than the alternative of a targeted ban for abuse. 

My GTC network requests were naive/brute force implementations. So if you add rate limiting, I'll do a followup task to see if there are any optimizations to be found. I vaguely recall forgoing some cache opportunities in favor of getting the functionality finished.

Reply:
I haven't had any issues with GoigToCamp - so far the only issue has been with some of the new recreation.gov Daily providers where their Cloudfront CDN blocked my IP temporarily for too many requests. 

My thinking around it is that `camply` can only check once every 5 minutes using the most frequent possible polling interval - so if the task takes 18s vs 9s it's not the worst thing in the world. But I obviously want to be careful when doing anything that will decrease performance - while also making sure we are being good consumers of the providers' APIs.

I noticed there's a lot of requests going on when using GoingToCamp so I thought I might get ahead of the issue when it gets more use in the summer months:

```console
[2023-02-02 10:41:56] CAMPLY   camply, the campsite finder ⛺️
[2023-02-02 10:41:56] DEBUG    Setting up camply debugging
[2023-02-02 10:41:56] DEBUG    Camply Version: 0.9.0
[2023-02-02 10:41:56] DEBUG    Python Version: 3.10.9
[2023-02-02 10:41:56] DEBUG    Platform: darwin
[2023-02-02 10:41:56] INFO     1 booking nights selected for search, ranging from 2023-09-01 to 2023-09-01
[2023-02-02 10:41:56] INFO     Retrieving Facility Information for Recreation Area ID: `4`.
[2023-02-02 10:41:56] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:57] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resourcelocation/rootmaps HTTP/1.1" 200 None
[2023-02-02 10:41:57] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:57] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resourcelocation HTTP/1.1" 200 None
[2023-02-02 10:41:57] INFO     1 Matching Campgrounds Found
[2023-02-02 10:41:57] INFO     ⛰  Washington State Parks, Washington (#4) - 🏕  Bay View State Park (#-2147483645)
[2023-02-02 10:41:57] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:57] DEBUG    https://washington.goingtocamp.com:443 "GET
                               /api/availability/map?mapId=-2147483394&resourceLocationId=-2147483645&bookingCategoryId=0&startDate=2023-09-01&endDate=2023-09-02&isReserving=True&getDailyAvailability=False&partySize=1&equipmentCategoryId=-32768 HTTP/1.1" 200 None
[2023-02-02 10:41:57] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:58] DEBUG    https://washington.goingtocamp.com:443 "GET
                               /api/availability/map?mapId=-2147483463&resourceLocationId=-2147483645&bookingCategoryId=0&startDate=2023-09-01&endDate=2023-09-02&isReserving=True&getDailyAvailability=False&partySize=1&equipmentCategoryId=-32768 HTTP/1.1" 200 None
[2023-02-02 10:41:58] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:58] DEBUG    https://washington.goingtocamp.com:443 "GET
                               /api/availability/map?mapId=-2147483331&resourceLocationId=-2147483645&bookingCategoryId=0&startDate=2023-09-01&endDate=2023-09-02&isReserving=True&getDailyAvailability=False&partySize=1&equipmentCategoryId=-32768 HTTP/1.1" 200 None
[2023-02-02 10:41:58] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:58] DEBUG    https://washington.goingtocamp.com:443 "GET
                               /api/availability/map?mapId=-2147483330&resourceLocationId=-2147483645&bookingCategoryId=0&startDate=2023-09-01&endDate=2023-09-02&isReserving=True&getDailyAvailability=False&partySize=1&equipmentCategoryId=-32768 HTTP/1.1" 200 None
[2023-02-02 10:41:58] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:59] DEBUG    https://washington.goingtocamp.com:443 "GET /api/attribute/filterable HTTP/1.1" 200 None
[2023-02-02 10:41:59] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:59] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482820 HTTP/1.1" 200 None
[2023-02-02 10:41:59] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:59] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482819 HTTP/1.1" 200 None
[2023-02-02 10:41:59] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:41:59] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482818 HTTP/1.1" 200 None
[2023-02-02 10:41:59] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:00] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482817 HTTP/1.1" 200 None
[2023-02-02 10:42:00] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:00] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482816 HTTP/1.1" 200 None
[2023-02-02 10:42:00] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:00] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482815 HTTP/1.1" 200 None
[2023-02-02 10:42:00] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:00] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482814 HTTP/1.1" 200 None
[2023-02-02 10:42:00] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:00] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482812 HTTP/1.1" 200 None
[2023-02-02 10:42:00] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:01] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482811 HTTP/1.1" 200 None
[2023-02-02 10:42:01] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:01] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482810 HTTP/1.1" 200 None
[2023-02-02 10:42:01] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:02] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482807 HTTP/1.1" 200 None
[2023-02-02 10:42:02] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:02] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482806 HTTP/1.1" 200 None
[2023-02-02 10:42:02] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:02] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482805 HTTP/1.1" 200 None
[2023-02-02 10:42:02] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:02] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482804 HTTP/1.1" 200 None
[2023-02-02 10:42:02] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:02] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482802 HTTP/1.1" 200 None
[2023-02-02 10:42:02] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:03] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482801 HTTP/1.1" 200 None
[2023-02-02 10:42:03] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:03] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482800 HTTP/1.1" 200 None
[2023-02-02 10:42:03] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:03] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482798 HTTP/1.1" 200 None
[2023-02-02 10:42:03] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:03] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482797 HTTP/1.1" 200 None
[2023-02-02 10:42:03] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:03] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482796 HTTP/1.1" 200 None
[2023-02-02 10:42:03] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:04] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482795 HTTP/1.1" 200 None
[2023-02-02 10:42:04] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:04] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482793 HTTP/1.1" 200 None
[2023-02-02 10:42:04] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:04] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482792 HTTP/1.1" 200 None
[2023-02-02 10:42:04] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:04] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482791 HTTP/1.1" 200 None
[2023-02-02 10:42:04] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:05] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482790 HTTP/1.1" 200 None
[2023-02-02 10:42:05] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:05] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482789 HTTP/1.1" 200 None
[2023-02-02 10:42:05] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:05] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482788 HTTP/1.1" 200 None
[2023-02-02 10:42:05] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:06] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482787 HTTP/1.1" 200 None
[2023-02-02 10:42:06] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:06] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482786 HTTP/1.1" 200 None
[2023-02-02 10:42:06] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:06] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482785 HTTP/1.1" 200 None
[2023-02-02 10:42:06] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:06] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482784 HTTP/1.1" 200 None
[2023-02-02 10:42:06] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:06] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482783 HTTP/1.1" 200 None
[2023-02-02 10:42:06] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:07] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482782 HTTP/1.1" 200 None
[2023-02-02 10:42:07] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:07] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482778 HTTP/1.1" 200 None
[2023-02-02 10:42:07] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:07] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482773 HTTP/1.1" 200 None
[2023-02-02 10:42:07] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:07] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482770 HTTP/1.1" 200 None
[2023-02-02 10:42:07] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:08] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482768 HTTP/1.1" 200 None
[2023-02-02 10:42:08] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:08] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482767 HTTP/1.1" 200 None
[2023-02-02 10:42:08] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:08] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482763 HTTP/1.1" 200 None
[2023-02-02 10:42:08] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:08] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482761 HTTP/1.1" 200 None
[2023-02-02 10:42:08] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:09] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482759 HTTP/1.1" 200 None
[2023-02-02 10:42:09] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:09] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482755 HTTP/1.1" 200 None
[2023-02-02 10:42:09] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:09] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482754 HTTP/1.1" 200 None
[2023-02-02 10:42:09] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:09] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482751 HTTP/1.1" 200 None
[2023-02-02 10:42:09] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:09] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482748 HTTP/1.1" 200 None
[2023-02-02 10:42:09] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:10] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482747 HTTP/1.1" 200 None
[2023-02-02 10:42:10] DEBUG    Starting new HTTPS connection (1): washington.goingtocamp.com:443
[2023-02-02 10:42:10] DEBUG    https://washington.goingtocamp.com:443 "GET /api/resource/details?resourceId=-2147482746 HTTP/1.1" 200 None
[2023-02-02 10:42:10] INFO     ⛺️ ⛺️ ⛺️ ⛺️ 47 Reservable Campsites Matching Search Preferences
[2023-02-02 10:42:10] INFO     📅 Fri, September 01 🏕  47 sites
[2023-02-02 10:42:10] INFO             ⛰️  Washington State Parks  🏕  Bay View State Park: ⛺ 47 sites
[2023-02-02 10:42:10] INFO                     🔗
                               https://washington.goingtocamp.com/create-booking/results?mapId=-2147483331&bookingCategoryId=0&startDate=2023-09-01&endDate=2023-09-02&isReserving=true&equipmentId=-32768&subEquipmentId=&partySize=1&resourceLocationId=-2147483645 (1
                               night)
[2023-02-02 10:42:10] CAMPLY   Exiting camply 👋
```

Reply:
A bit more details. 

Here is the likely crux of the problem: https://github.com/juftin/camply/blob/main/camply/search/search_going_to_camp.py#L147

In order to filter zero-occupancy sites, we have to fetch the site details of every eligible site to do the filtering. 

If you look at how `site_details` is used in the inner loop, you can see that it's mostly unimportant. It's used to: fill in the site occupancy, the site name, and the site type. I think when fetching availability, all of those data may be excluded without any loss of user functionality. But most importantly, it's used to filter out sites that can't be occupied. 

If we were to remove this call, the tradeoff is that we'd likely have to return zero-occupancy sites in the "available" site list, which is a pretty big tradeoff. That is, unless I made a mistake at the time about the API returning zero-occupancy sites regardless of request filters. We'd want to verify that this indeed does happen. If it does, it would degrade the user experience considerably, as it would mean users get results and notifications for sites that aren't viable. 

At the same time, removing occupancy filters would cut request counts by ~90%, so I think we should consider the refactor. 

I think the deciding factor should be how pervasive zero-occupancy sites are. I know at least one rec area uses this feature for pavilion/picnic area reservations. I believe that was Larrabee State Park in Washington. However, if zero-occupancy sites are the exception rather than the rule, removing the filter would greatly improve performance and reduce the likelihood of being rate limited.